JakinChan200/CS180_Roald
# Improve Country Performance

## Overview
Improves performance on country pages by 1 second.
Adds a country data point cache
Memoizes category results
Memoizes experimental result calculation for pub_trend

### To Do
Further optimizations could be done - couldn't implement lazy loading due to errors relating to our React version but is still a possibility to debug + implement

## Materials
- #58 and some of #60 


## Pictures
After removing the limit of 15 videos shown for general results and after moving sort up into the setStates after fetching from server:
<img width="794" alt="Screen Shot 2022-05-08 at 1 01 16 AM" src="https://user-images.githubusercontent.com/43893085/167287241-ac6b95ca-d8bd-46a8-be29-f25c52b9c1d8.png">

After memoizing the unoptimized experimental pub_time calculation and data fetching:
![image](https://user-images.githubusercontent.com/43893085/167992836-3b1ba256-e3c3-411f-8e5f-470711b16839.png)

After adding cache for general results:
![image](https://user-images.githubusercontent.com/43893085/167998017-fe7c60fe-9748-4eb2-8e22-fc08ecad039a.png)

After memoizing category results: 
![image](https://user-images.githubusercontent.com/43893085/168007938-60567833-269a-4973-9dd9-699da9230e3f.png)

After removing console.log statements:
![image](https://user-images.githubusercontent.com/43893085/168011778-3b59de59-4fe2-4999-841e-3578327c4751.png)

## Manual Testing Guide
- In `main` branch, navigate to a country component and record how long it takes to render
- Switch to `gobr002-improve-country` and navigate to the same country
- Note that the time taken to render each time is around a second less, and that the number of renders is fewer on each country

Pull requests should receive approval before merge.
